content/posts/0006-build-plan-buildpack.md

```diff
@@ -0,0 +1,137 @@
+[`bundler`](https://github.com/paketo-buildpacks/bundler) buildpacks provide
+those dependencies, and the [`puma`
+buildpack](https://github.com/paketo-buildpacks/puma) requires `gems` so that
+it can set the `puma` start command. With those buildpacks we now have a
+complete set of providing and requiring plan entries for `mri`, `bundler`, and
+`gems`. Now that all of the provisions and requirements in this buildpack group
+are satisfied, that group is chosen to execute its build phase.
+
+As you can see, the Build Plan is a fundamental part of the buildpack process.
+Its what allows buildpacks to be made modular, replaceable, and composable to
+enable a large number of use-cases. Its also what allows buildpacks to
+communicate and collaborate on what exactly gets done during the build phase.
+We'll explain more about that communication and collaboration a bit later. For
+now, let's focus on what the Build Plan Buildpack does.
+
+### How does the Build Plan Buildpack work?
+
+The Build Plan Buildpack works by reading a user provided file at the root of
+your app called `plan.toml`, which mirrors the [requires portion of Build
+Plan](https://github.com/buildpacks/spec/blob/main/buildpack.md#build-plan-toml).
+`plan.toml` has the following schema:
+```
+[[requires]]
+name = "<dependency name>"
+
+[requires.metadata]
+# buildpack-specific data
+
+[[or]]
+
+[[or.requires]]
+name = "<dependency name>"
+
+[or.requires.metadata]
+# buildpack-specific data
+```
+The buildpack reads `plan.toml` and makes all of the requirements listed in it.
+This means you will be able to make any arbitrary requirements that you want.
+Furthermore, if you are using the Build Plan Buildpack with Paketo Buildpacks,
+it will allow you to specify when you want the dependency to be present whether
+that is during build-time, launch-time, or both. You can specify this for
+Paketo Buildpacks because they respect two flags that go in the `metadata`
+field of a requirement. Those two flags are `build` which makes the dependency
+available during the build phase of subsequent buildpacks in the order, and
+`launch` which makes the dependency available in the final running image.
+
+### Why would I want this?
+
+Let's start with a basic example. Let's say that you want create an image that
+has the Go compiler on it. This is noramlly impossible because the `go`
+dependency is usually only ever required during build to compile your
+applications source code. However, with the Build Plan buildpack you could
+write the following `plan.toml` that would force the compiler to be on the
+image.
+```
+[[requires]]
+  name = "go"
+
+  [requires.metadata]
+    launch = true
+```
+
+There are other applications where the use of the Build Plan buildpack would
+enable a build that would be hard if not impossible to detect automatically, or
+to make an image as small as possible.
+
+An example of this would be an app that has a simple Nodejs frontend and has a Go backend.
```

Comment:
Why would you need node engine at build? This example is sort of vague enough to be confusing. Making this be a runnnable sample app would be awesome.

Reply:
Alternately, if there's an existing runnable example somewhere, adding it to the Paketo samples repo, and referencing it here would be great.

Reply:
This example needs to be overhauled.